src/libraries/System.IO.FileSystem/tests/FileStream/DeleteOnClose.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.IO.Tests
+{
+    public class FileStream_DeleteOnClose : FileSystemTest
+    {
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsFileLockingEnabled))]
+        public async Task DeleteOnClose_UsableAsMutex()
+        {
+            var cts = new CancellationTokenSource();
+            int enterCount = 0;
+            int locksRemaining = int.MaxValue;
+            bool exclusive = true;
+
+            string path = GetTestFilePath();
+            Assert.False(File.Exists(path));
+
+            Func<Task> lockFile = async () =>
+            {
+                while (!cts.IsCancellationRequested)
+                {
+                    try
+                    {
+                        using (var fs = new FileStream(path, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.None, 4096, FileOptions.DeleteOnClose))
+                        {
+                            int counter = Interlocked.Increment(ref enterCount);
+                            if (counter != 1)
+                            {
+                                exclusive = false;
+                                cts.Cancel();
+                                return;
+                            }
+
+                            // Hold the lock for a little bit.
+                            await Task.Delay(TimeSpan.FromMilliseconds(5));
+
+                            Interlocked.Decrement(ref enterCount);
+
+                            if (Interlocked.Decrement(ref locksRemaining) <= 0)
+                            {
+                                return;
+                            }
+                        }
+                        await Task.Delay(TimeSpan.FromMilliseconds(1));
+                    }
+                    catch (UnauthorizedAccessException)
+                    {
+                        // This can occur when the file is being deleted on Windows.
+                        await Task.Delay(TimeSpan.FromMilliseconds(1));
+                    }
+                    catch (IOException)
+                    {
+                        await Task.Delay(TimeSpan.FromMilliseconds(1));
+                    }
+                }
+            };
+
+            var tasks = new Task[50];
+            for (int i = 0; i < tasks.Length; i++)
+            {
+                tasks[i] = Task.Run(lockFile);
+            }
+
+            // Wait for 1000 locks.
+            cts.CancelAfter(TimeSpan.FromSeconds(100));
```

Comment:
Does it mean that it's always going to execute for at least 100 seconds? Can we decrease the value or move it to OuterLoop?

Reply:
As soon as we've obtained a 1000 locks, the test completes.
If it doesn't obtain a 1000 locks in 100 seconds, it cancels.

Reply:
Still, 100 seconds is a lot of time to wait in the case we can't reserve all those locks. I also agree with Adam: either reduce the number or move to OuterLoop.

Or maybe reduce the number of expected locks to reserve? Why does it have to be 1000? Would it be a source of failures in our test suite?

Reply:
This test only takes a couple of seconds. I picked a high timeout of 100 seconds to make sure it didn't timeout in CI. I've lowered the value now to 30 seconds.